File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        open_set = {"(", "{", "["}
        for i in s:
            if i in open_set:
                st.append(i)
            else:
                if not st:
                    return False
                o = st.pop()
                if i == ")" and o == "(":
                    continue
                elif i == "}" and o == "{":
                    continue
                elif i == "]" and o == "[":
                    continue
                else:
                    return False
        return not st
        

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> v;
        unordered_set<char> open = {'(', '{', '['};
        unordered_map<char, char> m = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };
        for(auto x: s) {
            if (open.count(x)) {
                v.push_back(x);
            } else {
                if (v.empty() || v.back() != m[x]) {
                    return false;
                } else {
                    v.pop_back();
                }
            }
        }
        return v.empty();
    }
};